# Import packages

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px

from functions_natalia import *

# Load dataset Demo.txt

In [2]:
df = pd.read_csv('../data/raw/df_final_demo.txt')

# Display the first few rows of the DataFrame
df.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   

         bal  calls_6_mnth  logons_6_mnth  
0   45105.30           6.0            9.0  
1  110860.30           6.0            9.0  
2   52467.79           6.0            9.0  
3   67454.65           3.0            6.0  
4  103671.75           0.0            3.0

# Perform initial data checking for Demo.txt

In [3]:
# Perform initial data checking for Demo.txt:
initial_data_checking(df)


Shape of the DataFrame:

(70609, 9)

Duplicate Rows Number:

0

Summary Statistics:

          client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  7.060900e+04    70595.000000      70595.000000  70594.000000   
mean   5.004992e+06       12.052950        150.659367     46.442240   
std    2.877278e+06        6.871819         82.089854     15.591273   
min    1.690000e+02        2.000000         33.000000     13.500000   
25%    2.519329e+06        6.000000         82.000000     32.500000   
50%    5.016978e+06       11.000000        136.000000     47.000000   
75%    7.483085e+06       16.000000        192.000000     59.000000   
max    9.999839e+06       62.000000        749.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  70595.000000  7.059500e+04  70595.000000   70595.000000  
mean       2.255528  1.474452e+05      3.382478       5.566740  
std        0.534997  3.015087e+05      2.236580       2.353286  
min        1.0

In [4]:
# Check Unique values, %, Missing values, %, data type:
unique_and_missing_values_dtype(df)

non-Null_count    dtype  unique_values %_unique  \
client_id                  70609    int64          70609   100.0%   
clnt_tenure_yr             70595  float64             54    0.08%   
clnt_tenure_mnth           70595  float64            482    0.68%   
clnt_age                   70594  float64            165    0.23%   
gendr                      70595   object              4    0.01%   
num_accts                  70595  float64              8    0.01%   
bal                        70595  float64          70328    99.6%   
calls_6_mnth               70595  float64              8    0.01%   
logons_6_mnth              70595  float64              9    0.01%   

                  missing_values %_missing  
client_id                      0      0.0%  
clnt_tenure_yr                14     0.02%  
clnt_tenure_mnth              14     0.02%  
clnt_age                      15     0.02%  
gendr                         14     0.02%  
num_accts                     14     0.02%  
bal                           14     0.02%  
calls_6_mnth                  14     0.02%  
logons_6_mnth                 14     0.02%

# Separate categorical and numerical columns for Demo.txt

In [5]:
# separate categorical and numerical columns:
separate_columns(df)


categorical_cols:

Index(['gendr'], dtype='object')

numerical_cols:

Index(['client_id', 'clnt_tenure_yr', 'clnt_tenure_mnth', 'clnt_age',
       'num_accts', 'bal', 'calls_6_mnth', 'logons_6_mnth'],
      dtype='object')


In [6]:
# analyze_numerical cols:
analyze_numerical(df) 


Numerical Columns Analysis:


client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  7.060900e+04    70595.000000      70595.000000  70594.000000   
mean   5.004992e+06       12.052950        150.659367     46.442240   
std    2.877278e+06        6.871819         82.089854     15.591273   
min    1.690000e+02        2.000000         33.000000     13.500000   
25%    2.519329e+06        6.000000         82.000000     32.500000   
50%    5.016978e+06       11.000000        136.000000     47.000000   
75%    7.483085e+06       16.000000        192.000000     59.000000   
max    9.999839e+06       62.000000        749.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  70595.000000  7.059500e+04  70595.000000   70595.000000  
mean       2.255528  1.474452e+05      3.382478       5.566740  
std        0.534997  3.015087e+05      2.236580       2.353286  
min        1.000000  1.378942e+04      0.000000       1.000000  
25%        2.000000  3.734683e+04      1.000000       4.000000  
50%        2.000000  6.333290e+04      3.000000       5.000000  
75%        2.000000  1.375449e+05      6.000000       7.000000  
max        8.000000  1.632004e+07      7.000000       9.000000

In [7]:
# analyze_categorical cols:
analyze_categorical(df) 

gendr
count   70595
unique      4
top         U
freq    24122

In [8]:
gender = df['gendr'].value_counts()
gender

gendr
U    24122
M    23724
F    22746
X        3
Name: count, dtype: int64

# Filling missing data (numerical_data with 'mean',categorical_data with 'mode'

In [9]:
# script for filling missing data (numerical_data with 'mean',categorical_data with 'mode' :

fill_missing_values(df)

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

# Load dataset Experiment_clients

In [15]:
df2 = pd.read_csv('../data/raw/df_final_experiment_clients.txt')

# Display the first few rows of the DataFrame
df2.head()

client_id Variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control

In [11]:
# Check Unique values, %, Missing values, %, data type:
unique_and_missing_values_dtype(df2)

non-Null_count   dtype  unique_values %_unique  missing_values  \
client_id           70609   int64          70609   100.0%               0   
Variation           50500  object              2     0.0%           20109   

          %_missing  
client_id      0.0%  
Variation    28.48%

In [12]:
initial_data_checking(df2)


Shape of the DataFrame:

(70609, 2)

Duplicate Rows Number:

0

Summary Statistics:

          client_id
count  7.060900e+04
mean   5.004992e+06
std    2.877278e+06
min    1.690000e+02
25%    2.519329e+06
50%    5.016978e+06
75%    7.483085e+06
max    9.999839e+06


In [16]:
df2.columns

Index(['client_id', 'Variation'], dtype='object')

In [17]:
# Count the occurrences of each value in the 'Variation' column
variation = df2['Variation'].value_counts()

# Display the result
print(variation)

Variation
Test       26968
Control    23532
Name: count, dtype: int64


In [21]:
#drop missng vlus:
df2_cleaned = df2.dropna(subset=['Variation'])
df2_cleaned.nunique()

client_id    50500
Variation        2
dtype: int64

# Load Data Pt-1

In [23]:
df3 = pd.read_csv('../data/raw/df_final_web_data_pt_1.txt')

# Display the first few rows of the DataFrame
df3.head()

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

In [24]:
# Perform initial data checking :
initial_data_checking(df3)


Shape of the DataFrame:

(343141, 5)

Duplicate Rows Number:

2095

Summary Statistics:

          client_id
count  3.431410e+05
mean   4.996097e+06
std    2.875839e+06
min    1.690000e+02
25%    2.506067e+06
50%    5.007915e+06
75%    7.468519e+06
max    9.999839e+06


In [25]:
# Check Unique values, %, Missing values, %, data type:
unique_and_missing_values_dtype(df3)

non-Null_count   dtype  unique_values %_unique  missing_values  \
client_id             343141   int64          58391   17.02%               0   
visitor_id            343141  object          62936   18.34%               0   
visit_id              343141  object          75256   21.93%               0   
process_step          343141  object              5     0.0%               0   
date_time             343141  object         282666   82.38%               0   

             %_missing  
client_id         0.0%  
visitor_id        0.0%  
visit_id          0.0%  
process_step      0.0%  
date_time         0.0%

### delete duplicate values:

In [28]:
df3_cleaned = df3.drop_duplicates()

In [29]:
# Check Print the count of duplicate rows
print("\nDuplicate Rows Number:\n")
print(df3_cleaned.duplicated().sum())


Duplicate Rows Number:

0


# Load Data Pt-2

In [32]:
df4 = pd.read_csv('../data/raw/df_final_web_data_pt_2.txt')

# Display the first few rows of the DataFrame
df4.head()

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58

In [33]:
# Perform initial data checking for Demo.txt:
initial_data_checking(df4)


Shape of the DataFrame:

(412264, 5)

Duplicate Rows Number:

8669

Summary Statistics:

          client_id
count  4.122640e+05
mean   5.028227e+06
std    2.881828e+06
min    3.360000e+02
25%    2.533952e+06
50%    5.056314e+06
75%    7.516792e+06
max    9.999875e+06


In [34]:
# Check Unique values, %, Missing values, %, data type:
unique_and_missing_values_dtype(df4)

non-Null_count   dtype  unique_values %_unique  missing_values  \
client_id             412264   int64          67430   16.36%               0   
visitor_id            412264  object          71042   17.23%               0   
visit_id              412264  object          82841   20.09%               0   
process_step          412264  object              5     0.0%               0   
date_time             412264  object         346697    84.1%               0   

             %_missing  
client_id         0.0%  
visitor_id        0.0%  
visit_id          0.0%  
process_step      0.0%  
date_time         0.0%

In [36]:
# filter duplicated rows

# Find duplicated rows
duplicates4 = df4.duplicated(keep=False)  # 'keep=False' marks all duplicates as True

# Display only the duplicated rows
df4_duplicates_only = df4[duplicates4]

print(df4_duplicates_only)

        client_id             visitor_id                      visit_id  \
124       3751049  937925261_74821699083  628817270_45049432060_475841   
125       3751049  937925261_74821699083  628817270_45049432060_475841   
142       9912156  746623513_52101087617  780044921_57678038107_984942   
143       9912156  746623513_52101087617  780044921_57678038107_984942   
325       5786602  893139880_61416949586  747153834_69434320996_858907   
...           ...                    ...                           ...   
412165    4408245   14058946_73585431797  994148957_84373994761_991865   
412166    4408245   14058946_73585431797  994148957_84373994761_991865   
412167    4408245   14058946_73585431797  994148957_84373994761_991865   
412216    3897324   80177429_24869309091  149796456_64325617570_112561   
412217    3897324   80177429_24869309091  149796456_64325617570_112561   

       process_step            date_time  
124           start  2017-05-15 23:09:44  
125           start  2017

In [37]:
df4_cleaned = df4.drop_duplicates()
# Check Print the count of duplicate rows
print("\nDuplicate Rows Number:\n")
print(df4_cleaned.duplicated().sum())


Duplicate Rows Number:

0


# Summary:
1. df_final_demo.txt = 15 missing value. Solution: filling missing data numerical_data with 'mean',categorical_data with 'mode'
2. df_final_experiment_clients.txt  = 20109 missing values. Solution : drop()
3. df_final_web_data_pt_1.txt  = Duplicate Rows Number: 2095 . Solution ? 
3. df_final_web_data_pt_2.txt  = Duplicate Rows Number: 8669 . Solution ? 


## Concat tables

In [38]:
df_final = pd.concat([df3_cleaned, df4_cleaned ], axis = 0)
df_final

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
412259    9668240   388766751_9038881013   922267647_3096648104_968866   
412260    9668240   388766751_9038881013   922267647_3096648104_968866   
412261    9668240   388766751_9038881013   922267647_3096648104_968866   
412262    9668240   388766751_9038881013   922267647_3096648104_968866   
412263     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
412259        start  2017-05-24 18:46:10  
412260        start  2017-05-24 18:45:29  
412261       step_1  2017-05-24 18:44:51  
412262        start  2017-05-24 18:44:34  
412263        start  2017-06-03 12:17:09  

[744641 rows x 5 columns]

In [41]:
   # Print the count of duplicate rows
print("\nDuplicate Rows Number:\n")
print(df_final.duplicated().sum())


Duplicate Rows Number:

0


In [42]:
initial_data_checking(df_final)


Shape of the DataFrame:

(744641, 5)

Duplicate Rows Number:

0

Summary Statistics:

          client_id
count  7.446410e+05
mean   5.014059e+06
std    2.879588e+06
min    1.690000e+02
25%    2.521851e+06
50%    5.031965e+06
75%    7.497388e+06
max    9.999875e+06
